java-frontend/src/main/java/org/sonar/java/se/symbolicvalues/RelationalSymbolicValue.java

```diff
@@ -209,42 +198,202 @@ RelationalSymbolicValue inverse() {
+      return RelationStateTable.solveRelation(kind, relation.kind);
+    }
+    return RelationState.UNDETERMINED;
+  }
+
+  private List<RelationalSymbolicValue> transitiveRelations(ProgramState programState) {
+    return knownRelations(programState)
+      .map(this::deduceTransitiveOrSimplified)
+      .filter(Objects::nonNull)
+      .collect(Collectors.toList());
+  }
+
+  private static Stream<RelationalSymbolicValue> knownRelations(ProgramState programState) {
+    return programState.getValuesWithConstraints(BooleanConstraint.TRUE)
+      .stream()
+      .filter(sv -> sv instanceof RelationalSymbolicValue)
+      .map(sv -> (RelationalSymbolicValue) sv);
+  }
+
+  @VisibleForTesting
+  RelationalSymbolicValue deduceTransitiveOrSimplified(RelationalSymbolicValue other) {
+    RelationalSymbolicValue result = simplify(other);
+    if (result != null) {
+      return result;
+    }
+    return combineTransitively(other);
+  }
+
+  @CheckForNull
+  private RelationalSymbolicValue simplify(RelationalSymbolicValue other) {
+    // a >= b && b >= a -> a == b
+    if (kind == GREATER_THAN_OR_EQUAL && other.kind == GREATER_THAN_OR_EQUAL
+      && hasSameOperandsAs(other) && !equals(other)) {
+      return new RelationalSymbolicValue(EQUAL, leftOp, rightOp);
+    }
+    return null;
+  }
+
+  @VisibleForTesting
+  boolean potentiallyTransitiveWith(RelationalSymbolicValue other) {
+    if (hasSameOperand() || other.hasSameOperand()) {
+      return false;
     }
-    return binaryRelation;
+    return (hasOperand(other.leftOp) || hasOperand(other.rightOp)) && !hasSameOperandsAs(other);
+  }
+
+  @CheckForNull
+  private RelationalSymbolicValue combineTransitively(RelationalSymbolicValue other) {
+    if (!potentiallyTransitiveWith(other)) {
+      return null;
+    }
+    RelationalSymbolicValue transitive = combineTransitivelyOneWay(other);
+    if (transitive != null) {
+      return transitive;
+    }
+    return other.combineTransitivelyOneWay(this);
+  }
+
+  @CheckForNull
+  private RelationalSymbolicValue combineTransitivelyOneWay(RelationalSymbolicValue other) {
+    RelationalSymbolicValue transitive = equalityTransitiveBuilder(other);
+    if (transitive != null) {
+      return transitive;
+    }
+    transitive = lessThanTransitiveBuilder(other);
+    if (transitive != null) {
+      return transitive;
+    }
+    return greaterThanEqualTransitiveBuilder(other);
+  }
+
+  @CheckForNull
+  private RelationalSymbolicValue equalityTransitiveBuilder(RelationalSymbolicValue other) {
+    if (!isEquality()
+      || (kind == METHOD_EQUALS && other.kind == EQUAL)) {
+      return null;
+    }
+
+    return new RelationalSymbolicValue(other.kind,
+      hasOperand(other.leftOp) ? differentOperand(other) : other.leftOp,
+      hasOperand(other.leftOp) ? other.rightOp : differentOperand(other));
+  }
+
+  @CheckForNull
+  private RelationalSymbolicValue lessThanTransitiveBuilder(RelationalSymbolicValue other) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce the number of returns of this method 6, down to the maximum allowed 5. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1142)